### PR TITLE
koka-2.3.9 (new formula)

### PR DESCRIPTION
**Pull request by daanx**
This adds a formula for the Koka compiler and interpreter.
See <http://koka-lang.org> for more information.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

note: audit complains that the default branch of head should be "master", but in this case it should be "dev". 

Thanks!

### CODE REVIEW

File: Formula/koka.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  def install
+    system "cabal", "new-update"
+    system "cabal", "new-build", "-j"
+    system "cabal", "new-run", "koka", "--",
+           "-e", "util/bundle.kk", "--",
+           "--prefix=bundle/local-brew", "--postfix=brew"
+    share.install Dir["bundle/local-brew/share/*"]
+    lib.install Dir["bundle/local-brew/lib/*"]
+    bin.install Dir["bundle/local-brew/bin/*"]
```
Comment (carlocab) [inline Formula/koka.rb:26]: Why not
```suggestion
           "--prefix=#{prefix}", "--postfix=brew"
```
?

File: Formula/koka.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  homepage "http://koka-lang.org"
+  url "https://github.com/koka-lang/koka/releases/download/v2.3.8/koka-v2.3.9-source.tar.gz"
+  sha256 "ae3e8127b46e51cc4df3525c5f913ecfbaeeb29f0a268e4a3d2e0ac809bc987f"
+  license "Apache-2.0"
+  head "https://github.com/koka-lang/koka.git", branch: "dev"
+
+  livecheck do
+    url :stable
+    strategy :github_latest
+  end
```
Comment (carlocab) [inline Formula/koka.rb:12]: Why is this needed?

File: Formula/koka.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        $ curl -sSL https://github.com/koka-lang/koka/releases/latest/download/install.sh | sh -s -- --uninstall --version=v#{kokaver}
+
+        or by only removing the symlink (to '/usr/local/bin/koka-v#{kokaver}'):
+
+        $ sudo rm /usr/local/bin/koka
+      EOS
+    else
+      ""
+    end
+  end
```
Comment (carlocab) [inline Formula/koka.rb:47]: This isn't really specific to Homebrew or they way we package something, which is what `caveats` are for. Let's leave this out.

File: Formula/koka.rb (hunk 1 above)
Comment (daanx) [reply Formula/koka.rb:26]: Thank you for the super fast reply :-) The reason is that the file times in share should be less than the ones in lib as it contains (koka) pre-compiled modules. If this is not the case, the koka compiler dependency analysis gets confused. 

To ensure this, I decided to explicitly copy the `share` files first, followed by the `lib` files.

File: Formula/koka.rb (hunk 2 above)
Comment (daanx) [reply Formula/koka.rb:12]: No idea :-) It's copied from another formula. Shall I remove it?

File: Formula/koka.rb (hunk 3 above)
Comment (daanx) [reply Formula/koka.rb:47]: I agree; but if a user had already installed koka before, and now starts using `brew`, then the new install will not work as the older koka will be found first.  This message will make this apparent and potentially save people from trying to figure out why the install is not working. Hope we can leave it in?

**Comment by daanx** [comment]
I see the CI fails due to the default branch being "dev"... is there a workaround?

File: Formula/koka.rb (hunk 1 above)
Comment (daanx) [reply Formula/koka.rb:26]: Also: the `--prefix=bundle/local-brew` flag is not related at all to the real install prefix but used by the koka bundler to create a correct local install tree from which we eventually install. It will not work correctly if we try to bundle directly into the real install prefix.  (I will rename that flag in the future to avoid the confusion)

File: Formula/koka.rb (hunk 2 above)
Comment (daanx) [reply Formula/koka.rb]: I just pushed an update that removes this.

**Comment by nandahkrishna** [comment]
> I see the CI fails due to the default branch being "dev"... is there a workaround?

You can get around this by creating `audit_exceptions/head_non_default_branch_allowlist.json` with the following contents:
```json
{
  "koka": "dev"
}
```

**Review by nandahkrishna — COMMENTED** [review]
It seems like the latest release (and its associated tag) mention `2.3.8`, but the source file contains, perhaps erroneously, the version number `2.3.9`. This can be seen in the `livecheck` output (or `brew info` too):
```
➜ brew lc koka
koka : 2.3.9 ==> 2.3.8
```

**Comment by daanx** [comment]
> > I see the CI fails due to the default branch being "dev"... is there a workaround?
> 
> You can get around this by creating `audit_exceptions/head_non_default_branch_allowlist.json` 

Ah, It looks I would be the first exception here -- that seems a bit much.  I think koka can use the `master` branch and I can change the development model for Koka to more regularly update master.

The version is actually correct: at this point the source bundle is v2.3.9 -- it is just that the release url happens to also contain a v2.3.8 part in its path. This is just bootstrapping -- once Koka v2.4.0 is released these will match going forward.

File: Formula/koka.rb (hunk 3 above)
Comment (alebcay) [reply Formula/koka.rb:42]: The installation will actually fail at the link stage if `/usr/local/bin/koka` already exists (Homebrew will not overwrite the file if it is already there). This message would not be displayed; Homebrew would display an error message instead.

Comment (daanx) [reply Formula/koka.rb:42]: ah I see; on my M1 brew installs to `/opt/homebrew/bin` but I see that is not always the case. I guess it's best to remove it. I'll add a note on the Koka install page instead.

Comment (alebcay) [reply Formula/koka.rb]: Just to clarify, users running Intel Macs would see the Homebrew error message since Homebrew would try to install to `/usr/local/bin/koka`; on M1 Macs Homebrew installs to `/opt/homebrew/bin/koka` so there would be no error message/conflict.

File: Formula/koka.rb (hunk 1 above)
Comment (daanx) [reply Formula/koka.rb]: I gave this more thought, and it turns out we can indeed bundle/install directly into the prefix; commit has been pushed.  I think the final formula looks quite good now -- much shorter :-)

File: Formula/koka.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  sha256 "ae3e8127b46e51cc4df3525c5f913ecfbaeeb29f0a268e4a3d2e0ac809bc987f"
+  license "Apache-2.0"
+  head "https://github.com/koka-lang/koka.git", branch: "master"
+
+  depends_on "cabal-install" => :build
+  depends_on "ghc" => :build
+  depends_on "vcpkg"
+
+  def install
+    system "cabal", "new-update"
```
Comment (SMillerDev) [inline Formula/koka.rb:14]: Does this pull in dependencies?

Comment (carlocab) [reply Formula/koka.rb:14]: We kinda do the same thing elsewhere with `cabal v2-update`. Though now I'm wondering what the difference is between `v2-update` and `new-update`.

Comment (alebcay) [reply Formula/koka.rb:14]: [I think they're the same](https://stackoverflow.com/questions/66535628/what-is-the-difference-between-install-v1-install-v2-install-and-new-install). TLDR: For modern versions of `cabal`, `v2-*` is the same as `*` or `new-*`. `v1-*` refers explicitly to the older deprecated commands.

Comment (carlocab) [reply Formula/koka.rb:14]: We should probably use `v2-*` instead of `new-*` to be consistent with other formulae then. We probably also want to use `std_cabal_v2_args`.

Comment (daanx) [reply Formula/koka.rb]: Thanks for the detailed feedback. The `new-` and `v2-` prefixes are equivalent, and since Cabal version 3 also equal to the plain `update`, `build`, and `run`. I updated to use the plain commands.

I am not using `std_cabal_v2_args` since those are meant for `cabal install` and not for `cabal build`. Koka is not using `cabal install` as it has its own bundler (since it also needs to include the standard libraries etc). 

I think it looks quite good now. Thanks.

Comment (carlocab) [reply Formula/koka.rb]: Thanks for the clarification. What does Koka bundle? Is that why you need `vcpkg`? More generally, what is the `vcpkg` dependency used for?

**Comment by daanx** [comment]
Since this is taking awhile, the new Koka version v2.4.0 has been released, and I updated the formula accordingly. (which is good because now in the url the directory and package version match.)

File: Formula/koka.rb (hunk 4 above)
Comment (daanx) [reply Formula/koka.rb]: First the Koka compiler is built (using `cabal build`) and then the compiler itself executes the `bundle.kk` script -- this script uses the newly build compiler to compile all standard libraries (in three variants (debug, release, etc)), then tar's everything together into a single tar archive, and 
finally installs the results (to <prefix>/bin <prefix>/lib and <prefix>/share).  

The `vcpkg` dependency is used when building the standard libraries for Koka; some of these use external C libraries and header files, like the `std/text/regex` module links with `pcre2`. When compiling the module Koka invokes `vcpkg` to install/resolve the required `pcre2` library+headers.

**Comment by daanx** [comment]
@carlocab: Is there anything more I should do to get this accepted? Thanks again.

File: Formula/koka.rb (hunk 4 above)
Comment (carlocab) [reply Formula/koka.rb]: When is the standard library built? Does that happen when this formula is built, or do users build it for themselves later on?

**Comment by carlocab** [comment]
Sorry, missed your response in a review thread. I have too many GitHub notifications.

**Comment by daanx** [comment]
> Sorry, missed your response in a review thread. I have too many GitHub notifications.

No worries -- thank you for all the work on Homebrew! :-)

File: Formula/koka.rb (hunk 4 above)
Comment (daanx) [reply Formula/koka.rb]: The standard libraries should be built when the formula is built. They are built in 3 variants and it takes quite a while so they are usually distributed with the compiler in binary form (so the compiler is snappy when loading the standard libraries).

Comment (carlocab) [reply Formula/koka.rb]: This is probably now broken on the tip of the `master` branch then, since the `vcpkg` formula now needs configuration to work properly.

In any case, we should be using Homebrew `pcre2` here instead of what's provided by `vcpkg`. Using `vcpkg`-provided `pcre2` sounds like it relies on state that won't survive repackaging. (i.e. `brew install koka` will result in a broken standard library unless you compiled it from source yourself).

Comment (daanx) [reply Formula/koka.rb]: Actually repackaging will will work fine as koka always links statically and includes the needed .o or .lib files so it will not break if reinstalled in binary form (and the end user needs to install no other libraries).

I thought about including all needed C library dependencies in the brew formula itself ... but that may be troublesome:

1. Currently the Koka modules themselves declare packages they rely on; I would need to ensure manually this matches what I put in the formula.
2. Currently, the Koka compiler knows about vcpkg and Conan -- if we install libraries through Brew we need to probably add some more support for that in the compiler.
3. Koka modules can declare the versions of C libraries they depend on (or what versions should not be used): this is supported by vcpkg and Conan, but not by Brew :-(
4. There are many more C libraries available through vcpkg and Conan than through Brew which seems to more used for end-user applications? 

I can potentially do 1 and 2 (but not ideal and would take some work), but 2 & 3 could prove problematic especially if the particular library is not supported by Brew... (however, it seems I only need pcre2, libuv, and libz so for Koka it may work out.) 

This is a general problem with tools like Brew, or apt, or pkg, etc. which seem geared to end-user applications while vcpkg and Conan are more oriented to C library installation. What is the end-goal of Brew here -- support all C libraries?

Comment (daanx) [reply Formula/koka.rb]: Ah apologies for the long reply. I guess tldr:

1. Koka has a build dependency on vcpkg (or Conan) to get the exact right versions of some C libraries needed for the standard libraries which copy archives and link statically
2. There is also a install dependency on vcpckg (or Conan) for the user since a user may write a Koka module that needs itself to be linked to some other C libraries.

Comment (carlocab) [reply Formula/koka.rb]: I'm fine with users relying on `vcpkg` at runtime to satisfy their library dependencies. For things we build here, I'd rather we use Homebrew-provided (or system-provided) libraries and link with them dynamically.

`zlib` is provided by macOS, and never breaks their ABI. `libuv` follows semver, and should be safe to link with dynamically. (Recent security issues also make me hesitant to link statically here.) Homebrew `pcre2` carries some macOS-specific fixes that are typically not found in other builds. (This is actually true more generally about Homebrew-built libraries, but I know we can only go so far in using them.)

**Comment by daanx** [comment]
Hi @carlocab ,  I updated to explicitly give the dependencies and not use `vcpkg` for that. I think the formula is good now. Can it be merged?

**Review by carlocab — COMMENTED** [review]
Apologies again for the delay. Just one last question.

File: Formula/koka.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  desc "Compiler for the Koka language"
+  homepage "http://koka-lang.org"
+  url "https://github.com/koka-lang/koka/releases/download/v2.4.0/koka-v2.4.0-source.tar.gz"
+  sha256 "3433ffe6b78d0bac9b25be2a1b3230610fb6e8b4a6a059d592278184f074ebb1"
+  license "Apache-2.0"
+  head "https://github.com/koka-lang/koka.git", branch: "master"
+
+  depends_on "cabal-install" => :build
+  depends_on "ghc" => :build
+  depends_on "pcre2" => :build
```
[the remark this answers is not in the document]
Comment (carlocab) [reply Formula/koka.rb:11]: Yes, I was referring to `pcre2`. Presumably this means that you are using the static `libpcre2.a`. We have a very strong preference in favour of dynamic linking of dependencies (as is typical for a package manager).

**Review by carlocab — APPROVED** [review]
Thanks for the contribution and your patience here, @daanx.
